src/Lecture4.hs

```diff
@@ -134,8 +136,27 @@ errors. We will simply return an optional result here.
 🕯 HINT: Use the 'readMaybe' function from the 'Text.Read' module.
 -}
 
+split :: (Char -> Bool) -> [Char] -> [[Char]]
+split p str = loop str
+   where loop str
+            | null rest = [first]
+            | otherwise = first : loop (tail rest)
+            where (first, rest) = break p str
+
+
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow str = 
+   do let parsedStr = split (== ',') str
+      [name, type', amount] <- case parsedStr of [x, y, z] -> Just [x, y, z]
+                                                 _ -> Nothing
+      name' <- case name of [] -> Nothing
+                            x  -> Just x
+      rowTradeType <- readMaybe type'
+      rowCost <- case readMaybe amount of Just x -> if x >= 0 then Just x else Nothing
+                                          whatever -> whatever
+      Just (Row name' rowTradeType rowCost)
```

Comment:
You have an excellent monadic implementation 🔝 

I haven't explained the following during lectures but there's a nice way to write this parsing syntax using `do`-notation, `Monad` instance for `Maybe` and the `MonadFail` desugaring trick 🧙🏻 

You can read more about this trick here:

* https://kowainik.github.io/posts/haskell-mini-patterns#monadfail-sugar

Your implementation is already pretty neat and concise! However, applying this programming pattern and a few helper functions, the `parseRow` function will look like this:

```suggestion
parseRow str = do
    [name, type', amount] <- Just $ split (== ',') str
    guard $ name /= ""
    rowTradeType <- readMaybe type'
    rowCost <- readMaybe amount
    guard $ rowCost >= 0
    pure $ Row name rowTradeType rowCost
```

Reply:
I see, but I don't understand how to define the helper function `guard` here. Could you explain more?

Reply:
Sorry, I wasn't too clear. You don't need to define `guard`, it's a standard function 🙂 

Let's look at the definition of `guard` to understand how it works:

```haskell
guard           :: (Alternative f) => Bool -> f ()
guard True      =  pure ()
guard False     =  empty
```

Here it uses two polymorphic functions:

1. `pure`: it comes from `Applicative` (it's the same as `return` from `Monad`)
2. `empty` comes from the `Alternative`

If we specialise these functions to `Maybe`, the implementation will be equivalent to the following:

```haskell
guard           :: Bool -> Maybe ()
guard True      =  Just ()
guard False     =  Nothing
```

So the function returns `Nothing` on `False` and `Just ()` on `True`.

<hr>

When you write the code like this:

```haskell
rowCost <- readMaybe amount
guard $ rowCost >= 0
smthElse
```

it's equivalent to the following (after desugaring `do`-notation):

```haskell
readMaybe amount >>= \rowCost ->
guard (rowCost >= 0) >>= \_ ->
smthElse
```

So you can see that `guard` will return either `Just ()` (where we ignore `()` because it's not interesting) or `Nothing`.

`>>=` for `Maybe` either calls the given function or returns `Nothing` when the left argument is `Nothing`. So, in some sense it short-circuits and stops the function when the condition returns `False`.

On other words, it guards the following code and makes sure that the condition is true below `guard` 🙂 

I hope it makes it clearer! 🤗

Reply:
I got it. Thank you for explaining again! 👍🏼